src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -75,45 +97,100 @@ public async Task<ProjectInfo> CreateProjectInfoAsync(ProjectId projectId, Check
+
+        await Task.WhenAll(attributesTask, textTask).ConfigureAwait(false);
+    }
+
+    public async Task<DocumentInfo> CreateDocumentInfoAsync(
+        DocumentId documentId, Checksum attributeChecksum, Checksum textChecksum, CancellationToken cancellationToken)
+    {
+        var attributes = await GetAssetAsync<DocumentInfo.DocumentAttributes>(new(AssetPathKind.DocumentAttributes, documentId), attributeChecksum, cancellationToken).ConfigureAwait(false);
+        var serializableSourceText = await GetAssetAsync<SerializableSourceText>(new(AssetPathKind.DocumentText, documentId), textChecksum, cancellationToken).ConfigureAwait(false);
 
         var text = await serializableSourceText.GetTextAsync(cancellationToken).ConfigureAwait(false);
         var textLoader = TextLoader.From(TextAndVersion.Create(text, VersionStamp.Create(), attributes.FilePath));
 
         // TODO: do we need version?
         return new DocumentInfo(attributes, textLoader, documentServiceProvider: null);
     }
+}
+
+internal static class AbstractAssetProviderExtensions
+{
+    public static Task GetAssetsAsync<TAsset>(
+        this AbstractAssetProvider assetProvider, AssetPath assetPath, HashSet<Checksum> checksums, CancellationToken cancellationToken)
+    {
+        return assetProvider.GetAssetsAsync<TAsset, VoidResult>(
+            assetPath, checksums, callback: null, arg: default, cancellationToken);
+    }
+
+    public static Task GetAssetsAsync<T>(
+        this AbstractAssetProvider assetProvider, AssetPath assetPath, ChecksumCollection checksums, CancellationToken cancellationToken)
+    {
+        return assetProvider.GetAssetsAsync<T, VoidResult>(
+            assetPath, checksums, callback: null, arg: default, cancellationToken);
+    }
+
+    public static async Task GetAssetsAsync<T, TArg>(
+        this AbstractAssetProvider assetProvider, AssetPath assetPath, ChecksumCollection checksums, Action<Checksum, T, TArg>? callback, TArg? arg, CancellationToken cancellationToken)
+    {
+        using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksumSet);
+#if NET
+        checksumSet.EnsureCapacity(checksums.Children.Length);
+#endif
+        checksumSet.AddAll(checksums.Children);
+
+        await assetProvider.GetAssetsAsync(assetPath, checksumSet, callback, arg, cancellationToken).ConfigureAwait(false);
+    }
 
-    public async Task<ImmutableArray<T>> GetAssetsAsync<T>(
-        AssetPath assetPath, ChecksumCollection checksums, CancellationToken cancellationToken) where T : class
+    public static async Task<ImmutableArray<T>> GetAssetsArrayAsync<T>(
+        this AbstractAssetProvider assetProvider, AssetPath assetPath, ChecksumCollection checksums, CancellationToken cancellationToken) where T : class
     {
         using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksumSet);
         checksumSet.AddAll(checksums.Children);
 
         using var _2 = ArrayBuilder<T>.GetInstance(checksumSet.Count, out var builder);
```

Comment:
This feels like another place that could warrant using ImmutableCollectionsMarshal, but maybe that's made more difficult by needing to populate the array in a callback.

Reply:
yup.  have updated all of these in followup pr.